modulemd/v2/modulemd-profile.c

```diff
@@ -46,6 +46,50 @@ enum
 static GParamSpec *properties[N_PROPS];
 
 
+gboolean
+modulemd_profile_equals (ModulemdProfile *self, ModulemdProfile *other)
+{
+  gchar ** whitelist_self, ** whitelist_other;
+  GHashTableIter iter_1, iter_2;
+  gpointer key_1, key_2, value_1, value_2;
+
+  g_return_val_if_fail (MODULEMD_IS_BUILDOPTS (self), NULL);
+  g_return_val_if_fail (MODULEMD_IS_BUILDOPTS (other), NULL);
+
+  const gchar * name_self = modulemd_profile_get_name (self);
+  const gchar * name_other = modulemd_profile_get_name (other);
+  if(g_strcmp0 (name_self, name_other))
+    return FALSE;
+
+  const gchar * description_self = modulemd_profile_get_description (self, NULL);
+  const gchar * description_other = modulemd_profile_get_description (other, NULL);
+  if(g_strcmp0 (description_self, description_other))
+    return FALSE;
+
+  rpms_self = modulemd_profile_get_rpms_as_strv (self);
+  rpms_other = modulemd_profile_get_rpms_as_strv (other);
+
+  if (g_hash_table_size (rpms_self) != g_hash_table_size(rpms_other))
+    {
+      return FALSE;
+    }
+
+  g_hash_table_iter_init (&iter_1, rpms_self);
+  g_hash_table_iter_init (&iter_2, rpms_other);
+
+  while (g_hash_table_iter_next (&iter_1, &key_1, &value_1) && g_hash_table_iter_next (&iter_2, &key_2, &value_2))
```

Comment:
This iteration isn't going to work, in part because there is no guarantee that the keys are in the same order in each hash table. You probably need to first get the lists of keys (there's a utility function for this in modulemd-util.h) and compare those. Then, if they are identical, look up each value one at a time by the keys and compare them.